File: frontend/src/lib/schemas.ts
import { nullable, z } from "zod";

const N420Schema = z.object({
    leagueId: z.string(),
    queueType: z.string(),
    tier: z.string(),
    rank: z.string(),
    leaguePoints: z.number(),
    wins: z.number(),
    losses: z.number(),
    veteran: z.boolean(),
    inactive: z.boolean(),
    freshBlood: z.boolean(),
    hotStreak: z.boolean(),
});

const N440Schema = z.object({
    leagueId: z.string(),
    queueType: z.string(),
    tier: z.string(),
    rank: z.string(),
    leaguePoints: z.number(),
    wins: z.number(),
    losses: z.number(),
    veteran: z.boolean(),
    inactive: z.boolean(),
    freshBlood: z.boolean(),
    hotStreak: z.boolean(),
});

const LeagueEntriesSchema = z.object({
    "420": N420Schema.optional().nullable(),
    "440": N440Schema.optional().nullable(),
});

export const summonerSchema = z.object({
    gameName: z.string(),
    tagLine: z.string(),
    summonerId: z.string(),
    puuid: z.string(),
    name: z.string(),
    profileIconId: z.number(),
    summonerLevel: z.number(),
    platform: z.string(),
    leagueEntries: LeagueEntriesSchema,
});

// Infer the Summoner type from the schema
export type Summoner = z.infer<typeof summonerSchema>;

File: api/schemas.py
from enum import Enum
from pydantic import BaseModel


class TierEnum(str, Enum):
    CHALLENGER = 'CHALLENGER'
    GRANDMASTER = 'GRANDMASTER'
    MASTER = 'MASTER'
    DIAMOND = 'DIAMOND'
    EMERALD = 'EMERALD'
    PLATINUM = 'PLATINUM'
    GOLD = 'GOLD'
    SILVER = 'SILVER'
    BRONZE = 'BRONZE'
    IRON = 'IRON'

class RankEnum(str, Enum):
    I = 'I'
    II = 'II'
    III = 'III'
    IV = 'IV'


# Define the schema
class LeagueEntry(BaseModel):
    leagueId: str
    queueType: str
    tier: TierEnum
    rank: RankEnum
    leaguePoints: int
    wins: int
    losses: int
    veteran: bool
    inactive: bool
    freshBlood: bool
    hotStreak: bool

class Summoner(BaseModel):
    gameName: str
    tagLine: str
    summonerId: str
    puuid: str
    name: str
    profileIconId: int
    summonerLevel: int
    platform: str
    leagueEntries: dict[str, LeagueEntry | None]


class SummonerCreate(BaseModel):
    gameName: str
    tagLine: str
    platform: str
